File: tests/sage_linear_adam.cpp
// Copyright 2020 Marcel Wagenländer

#include "adam.hpp"
#include "chunking.hpp"
#include "cuda_helper.hpp"
#include "helper.hpp"
#include "sage_linear.hpp"
#include "tensors.hpp"

#include "catch2/catch.hpp"
#include <string>


const std::string home = std::getenv("HOME");
const std::string dir_path = home + "/gpu_memory_reduction/alzheimer/data";
const std::string flickr_dir_path = dir_path + "/flickr";
const std::string test_dir_path = dir_path + "/tests";


int test_sage_linear_adam(Matrix<float> *input_self, Matrix<float> *input_neigh, Matrix<float> *in_gradients) {
    std::string path;
    float learning_rate = 0.003;

    CudaHelper cuda_helper;
    SageLinear sage_linear_layer(&cuda_helper, input_self->num_columns_, in_gradients->num_columns_, input_self->num_rows_);
    std::vector<Matrix<float> *> params = sage_linear_layer.get_parameters();
    Adam adam(&cuda_helper, learning_rate, params, sage_linear_layer.get_gradients());

    save_params(params);

    Matrix<float> *activations = sage_linear_layer.forward(input_self, input_neigh);

    save_npy_matrix(activations, test_dir_path + "/activations.npy");

    SageLinearGradients *gradients = sage_linear_layer.backward(in_gradients);

    save_grads(gradients, sage_linear_layer.get_gradients());

    adam.step();

    save_npy_matrix(params[0], test_dir_path + "/self_weight_updated.npy");
    save_npy_matrix(params[1], test_dir_path + "/self_bias_updated.npy");
    save_npy_matrix(params[2], test_dir_path + "/neigh_weight_updated.npy");
    save_npy_matrix(params[3], test_dir_path + "/neigh_bias_updated.npy");

    char command[] = "/home/ubuntu/gpu_memory_reduction/pytorch-venv/bin/python3 /home/ubuntu/gpu_memory_reduction/alzheimer/tests/sage_linear_adam.py";
    system(command);

    path = test_dir_path + "/value.npy";
    return read_return_value(path);
}

int test_sage_linear_adam_chunked(Matrix<float> *input_self, Matrix<float> *input_neigh, Matrix<float> *incoming_gradients, long chunk_size) {
    std::string path;
    float learning_rate = 0.003;
    CudaHelper cuda_helper;

    long num_nodes = input_self->num_rows_;
    long num_in_features = input_self->num_columns_;
    long num_out_features = incoming_gradients->num_columns_;

    long num_chunks = ceil((float) num_nodes / (float) chunk_size);
    std::vector<Matrix<float>> input_self_chunked(num_chunks);
    std::vector<Matrix<float>> input_neigh_chunked(num_chunks);
    std::vector<Matrix<float>> incoming_gradients_chunked(num_chunks);
    chunk_up(input_self, &input_self_chunked, chunk_size);
    chunk_up(input_neigh, &input_neigh_chunked, chunk_size);
    chunk_up(incoming_gradients, &incoming_gradients_chunked, chunk_size);


    SageLinearChunked sage_linear_layer(&cuda_helper, num_in_features, num_out_features, chunk_size, num_nodes);
    std::vector<Matrix<float> *> params = sage_linear_layer.get_parameters();
    Adam adam(&cuda_helper, learning_rate, params, sage_linear_layer.get_gradients());

    save_params(params);

    std::vector<Matrix<float>> *activations = sage_linear_layer.forward(&input_self_chunked, &input_neigh_chunked);

    Matrix<float> activations_one(num_nodes, num_out_features, false);
    stitch(activations, &activations_one);
    save_npy_matrix(&activations_one, test_dir_path + "/activations.npy");

    SageLinearGradientsChunked *gradients = sage_linear_layer.backward(&incoming_gradients_chunked);

    save_grads(gradients, sage_linear_layer.get_gradients(), num_nodes);

    adam.step();

    save_npy_matrix(params[0], test_dir_path + "/self_weight_updated.npy");
    save_npy_matrix(params[1], test_dir_path + "/self_bias_updated.npy");
    save_npy_matrix(params[2], test_dir_path + "/neigh_weight_updated.npy");
    save_npy_matrix(params[3], test_dir_path + "/neigh_bias_updated.npy");

    char command[] = "/home/ubuntu/gpu_memory_reduction/pytorch-venv/bin/python3 /home/ubuntu/gpu_memory_reduction/alzheimer/tests/sage_linear_adam.py";
    system(command);

    path = test_dir_path + "/value.npy";
    return read_return_value(path);
}

int test_sage_linear_adam_nans(Matrix<float> *input_self, Matrix<float> *input_neigh, Matrix<float> *in_gradients) {
    std::string path;
    float learning_rate = 0.003;
    bool has_nans = true;

    CudaHelper cuda_helper;
    SageLinear sage_linear_layer(&cuda_helper, input_self->num_columns_, in_gradients->num_columns_, input_self->num_rows_);
    std::vector<Matrix<float> *> parameters = sage_linear_layer.get_parameters();
    std::vector<Matrix<float> *> parameters_gradients = sage_linear_layer.get_gradients();
    Adam adam(&cuda_helper, learning_rate, parameters, parameters_gradients);

    long num_iterations = 3;
    for (long i = 0; i < num_iterations; ++i) {
        Matrix<float> *activations = sage_linear_layer.forward(input_self, input_neigh);

        has_nans = has_nans && check_nans(activations, "Activations");

        SageLinearGradients *gradients = sage_linear_layer.backward(in_gradients);

        has_nans = has_nans && check_nans(gradients->self_gradients, "Input self gradients");
        has_nans = has_nans && check_nans(gradients->neighbourhood_gradients, "Input neigh gradients");
        for (int i = 0; i < parameters_gradients.size(); ++i) {
            has_nans = has_nans && check_nans(parameters_gradients[i], "Parameter gradient " + std::to_string(i));
        }

        adam.step();

        for (int i = 0; i < parameters_gradients.size(); ++i) {
            has_nans = has_nans && check_nans(parameters[i], "Parameter " + std::to_string(i));
        }
    }

    return !has_nans;
}

int test_sage_linear_adam_nans_chunked(Matrix<float> *input_self, Matrix<float> *input_neigh, Matrix<float> *incoming_gradients, long chunk_size) {
    std::string path;
    float learning_rate = 0.003;
    bool has_nans = true;
    CudaHelper cuda_helper;

    long num_nodes = input_self->num_rows_;
    long num_in_features = input_self->num_columns_;
    long num_out_features = incoming_gradients->num_columns_;

    long num_chunks = ceil((float) num_nodes / (float) chunk_size);
    std::vector<Matrix<float>> input_self_chunked(num_chunks);
    std::vector<Matrix<float>> input_neigh_chunked(num_chunks);
    std::vector<Matrix<float>> incoming_gradients_chunked(num_chunks);
    chunk_up(input_self, &input_self_chunked, chunk_size);
    chunk_up(input_neigh, &input_neigh_chunked, chunk_size);
    chunk_up(incoming_gradients, &incoming_gradients_chunked, chunk_size);

    SageLinearChunked sage_linear_layer(&cuda_helper, num_in_features, num_out_features, chunk_size, num_nodes);
    std::vector<Matrix<float> *> parameters = sage_linear_layer.get_parameters();
    std::vector<Matrix<float> *> parameters_gradients = sage_linear_layer.get_gradients();
    Adam adam(&cuda_helper, learning_rate, parameters, parameters_gradients);

    long num_iterations = 3;
    for (long i = 0; i < num_iterations; ++i) {
        std::vector<Matrix<float>> *activations = sage_linear_layer.forward(&input_self_chunked, &input_neigh_chunked);

        has_nans = has_nans && check_nans(activations, "Activations");

        SageLinearGradientsChunked *gradients = sage_linear_layer.backward(&incoming_gradients_chunked);

        has_nans = has_nans && check_nans(gradients->self_gradients, "Input self gradients");
        has_nans = has_nans && check_nans(gradients->neighbourhood_gradients, "Input neigh gradients");
        for (int i = 0; i < parameters_gradients.size(); ++i) {
            has_nans = has_nans && check_nans(parameters_gradients[i], "Parameter gradient " + std::to_string(i));
        }

        adam.step();

        for (int i = 0; i < parameters_gradients.size(); ++i) {
            has_nans = has_nans && check_nans(parameters[i], "Parameter " + std::to_string(i));
        }
    }

    return !has_nans;
}


TEST_CASE("SageLinear and Adam", "[sagelinear][adam]") {
    int rows = 1 << 15;
    int columns = 1 << 9;
    int num_out_features = 1 << 8;

    Matrix<float> input_self(rows, columns, true);
    input_self.set_random_values();
    save_npy_matrix(&input_self, test_dir_path + "/input_self.npy");

    Matrix<float> input_neigh(rows, columns, true);
    input_neigh.set_random_values();
    save_npy_matrix(&input_neigh, test_dir_path + "/input_neigh.npy");

    Matrix<float> in_gradients(rows, num_out_features, true);
    in_gradients.set_random_values();
    save_npy_matrix(&in_gradients, test_dir_path + "/in_gradients.npy");

    CHECK(test_sage_linear_adam(&input_self, &input_neigh, &in_gradients));
}

TEST_CASE("SageLinear and Adam chunked", "[sagelinear][adam][chunked]") {
    int rows = 1 << 15;
    int columns = 1 << 9;
    int num_out_features = 1 << 8;

    Matrix<float> input_self(rows, columns, true);
    input_self.set_random_values();
    save_npy_matrix(&input_self, test_dir_path + "/input_self.npy");

    Matrix<float> input_neigh(rows, columns, true);
    input_neigh.set_random_values();
    save_npy_matrix(&input_neigh, test_dir_path + "/input_neigh.npy");

    Matrix<float> in_gradients(rows, num_out_features, true);
    in_gradients.set_random_values();
    save_npy_matrix(&in_gradients, test_dir_path + "/in_gradients.npy");

    CHECK(test_sage_linear_adam_chunked(&input_self, &input_neigh, &in_gradients, 1 << 15));
    CHECK(test_sage_linear_adam_chunked(&input_self, &input_neigh, &in_gradients, 1 << 12));
    CHECK(test_sage_linear_adam_chunked(&input_self, &input_neigh, &in_gradients, 1 << 8));
}

TEST_CASE("SageLinear and Adam, NaNs", "[sagelinear][adam][nan]") {
    int rows = 2449029;// products
    int columns = 100; // products
    int num_out_features = 1 << 8;

    Matrix<float> input_self(rows, columns, true);
    input_self.set_random_values();

    Matrix<float> input_neigh(rows, columns, true);
    input_neigh.set_random_values();

    Matrix<float> in_gradients(rows, num_out_features, true);
    in_gradients.set_random_values();

    CHECK(test_sage_linear_adam_nans(&input_self, &input_neigh, &in_gradients));
}

TEST_CASE("SageLinear and Adam, chunked, NaNs", "[sagelinear][adam][chunked][nan]") {
    int rows = 2449029;// products
    int columns = 100; // products
    int num_out_features = 1 << 8;

    Matrix<float> input_self(rows, columns, true);
    input_self.set_random_values();

    Matrix<float> input_neigh(rows, columns, true);
    input_neigh.set_random_values();

    Matrix<float> in_gradients(rows, num_out_features, true);
    in_gradients.set_random_values();

    CHECK(test_sage_linear_adam_nans_chunked(&input_self, &input_neigh, &in_gradients, 1 << 15));
    CHECK(test_sage_linear_adam_nans_chunked(&input_self, &input_neigh, &in_gradients, 1 << 12));
    CHECK(test_sage_linear_adam_nans_chunked(&input_self, &input_neigh, &in_gradients, 1 << 8));
}

File: tests/sage_linear_adam.py
import numpy as np
import os
import scipy.io
import scipy.sparse as sp
import torch
from helper import (write_equal, print_close_equal, to_torch,
                    check_close_equal)


def test_sage_linear_adam():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    all_close = 1.0

    home = os.getenv("HOME")
    dir_path = home + "/gpu_memory_reduction/alzheimer/data"
    flickr_dir_path = dir_path + "/flickr"
    test_dir_path = dir_path + "/tests"

    # FORWARD
    path = test_dir_path + "/input_self.npy"
    input_self = np.load(path)
    path = test_dir_path + "/input_neigh.npy"
    input_neigh = np.load(path)
    path = test_dir_path + "/in_gradients.npy"
    in_gradients = np.load(path)

    input_self_torch = to_torch(input_self)
    input_neigh_torch = to_torch(input_neigh)

    path = test_dir_path + "/self_weight.npy"
    self_weight = np.load(path)
    path = test_dir_path + "/self_bias.npy"
    self_bias = np.load(path)
    path = test_dir_path + "/neigh_weight.npy"
    neigh_weight = np.load(path)
    path = test_dir_path + "/neigh_bias.npy"
    neigh_bias = np.load(path)
    path = test_dir_path + "/activations.npy"
    activations = np.load(path)

    self_weight_torch = to_torch(self_weight)
    self_bias_torch = to_torch(self_bias)
    neigh_weight_torch = to_torch(neigh_weight)
    neigh_bias_torch = to_torch(neigh_bias)

    learning_rate = 0.003
    params = [self_weight_torch,
              self_bias_torch,
              neigh_weight_torch,
              neigh_bias_torch]
    optimiser = torch.optim.Adam(params, lr=learning_rate)
    optimiser.zero_grad()

    self_result_torch = torch.matmul(input_self_torch, self_weight_torch) + self_bias_torch.T
    neigh_result_torch = torch.matmul(input_neigh_torch, neigh_weight_torch) + neigh_bias_torch.T
    true_sage_linear_result_torch = self_result_torch + neigh_result_torch

    true_sage_linear_result = true_sage_linear_result_torch.detach().cpu().numpy()

    ratio_close, ratio_equal = check_close_equal(activations, true_sage_linear_result)
    all_close = all_close * ratio_close
    print_close_equal("SageLinear", ratio_close, ratio_equal)

    # BACKPROPAGATION
    in_gradients_torch = torch.from_numpy(in_gradients)
    in_gradients_torch = in_gradients_torch.to(device)
    true_sage_linear_result_torch.backward(in_gradients_torch)

    path = test_dir_path + "/self_grads.npy"
    self_grads = np.load(path)
    path = test_dir_path + "/neigh_grads.npy"
    neigh_grads = np.load(path)

    path = test_dir_path + "/self_weight_grads.npy"
    self_weight_grads = np.load(path)
    path = test_dir_path + "/self_bias_grads.npy"
    self_bias_grads = np.load(path)
    path = test_dir_path + "/neigh_weight_grads.npy"
    neigh_weight_grads = np.load(path)
    path = test_dir_path + "/neigh_bias_grads.npy"
    neigh_bias_grads = np.load(path)

    true_self_grads = input_self_torch.grad.cpu().numpy()
    true_neigh_grads = input_neigh_torch.grad.cpu().numpy()
    true_self_weight_grads = self_weight_torch.grad.cpu().numpy()
    true_self_bias_grads = self_bias_torch.grad.cpu().numpy()
    true_neigh_weight_grads = neigh_weight_torch.grad.cpu().numpy()
    true_neigh_bias_grads = neigh_bias_torch.grad.cpu().numpy()

    ratio_close, ratio_equal = check_close_equal(self_grads, true_self_grads)
    all_close = all_close * ratio_close
    print_close_equal("LinearSage self", ratio_close, ratio_equal)
    ratio_close, ratio_equal = check_close_equal(neigh_grads, true_neigh_grads)
    all_close = all_close * ratio_close
    print_close_equal("LinearSage neighbourhood", ratio_close, ratio_equal)
    ratio_close, ratio_equal = check_close_equal(self_weight_grads, true_self_weight_grads)
    all_close = all_close * ratio_close
    print_close_equal("LinearSage self weight", ratio_close, ratio_equal)
    ratio_close, ratio_equal = check_close_equal(self_bias_grads, true_self_bias_grads)
    all_close = all_close * ratio_close
    print_close_equal("LinearSage self bias", ratio_close, ratio_equal)
    ratio_close, ratio_equal = check_close_equal(neigh_weight_grads, true_neigh_weight_grads)
    all_close = all_close * ratio_close
    print_close_equal("LinearSage neighbourhood weight", ratio_close, ratio_equal)
    ratio_close, ratio_equal = check_close_equal(neigh_bias_grads, true_neigh_bias_grads)
    all_close = all_close * ratio_close
    print_close_equal("LinearSage neighbourhood bias", ratio_close, ratio_equal)

    # ADAM
    optimiser.step()

    path = test_dir_path + "/self_weight_updated.npy";
    self_weight_updated = np.load(path)
    path = test_dir_path + "/self_bias_updated.npy";
    self_bias_updated = np.load(path)
    path = test_dir_path + "/neigh_weight_updated.npy";
    neigh_weight_updated = np.load(path)
    path = test_dir_path + "/neigh_bias_updated.npy";
    neigh_bias_updated = np.load(path)

    true_self_weight_updated = self_weight_torch.detach().cpu().numpy()
    ratio_close, ratio_equal = check_close_equal(self_weight_updated, true_self_weight_updated)
    all_close = all_close * ratio_close
    print_close_equal("Adam self weight", ratio_close, ratio_equal)

    true_self_bias_updated = self_bias_torch.detach().cpu().numpy()
    ratio_close, ratio_equal = check_close_equal(self_bias_updated, true_self_bias_updated)
    all_close = all_close * ratio_close
    print_close_equal("Adam self bias", ratio_close, ratio_equal)

    true_neigh_weight_updated = neigh_weight_torch.detach().cpu().numpy()
    ratio_close, ratio_equal = check_close_equal(neigh_weight_updated, true_neigh_weight_updated)
    all_close = all_close * ratio_close
    print_close_equal("Adam neighbourhood weight", ratio_close, ratio_equal)

    true_neigh_bias_updated = neigh_bias_torch.detach().cpu().numpy()
    ratio_close, ratio_equal = check_close_equal(neigh_bias_updated, true_neigh_bias_updated)
    all_close = all_close * ratio_close
    print_close_equal("Adam neighbourhood bias", ratio_close, ratio_equal)

    path = test_dir_path + "/value.npy"
    write_equal(all_close, path)


def compare_adam():
    home = os.getenv("HOME")
    dir_path = home + "/gpu_memory_reduction/alzheimer/data"
    flickr_dir_path = dir_path + "/flickr"
    test_dir_path = dir_path + "/tests"
    all_equal = 1.0

    for i in range(4):
        path = test_dir_path + "/gradient_" + str(i) + ".npy"
        gradient = np.load(path)

        path = test_dir_path + "/gradient_chunked_" + str(i) + ".npy"
        gradient_chunked = np.load(path)

        ratio_close, ratio_equal = check_close_equal(gradient, gradient_chunked)
        all_equal = all_equal * ratio_equal
        print_close_equal("Gradient layer {}".format(i), ratio_close, ratio_equal)

        path = test_dir_path + "/weight_" + str(i) + ".npy"
        weight = np.load(path)

        path = test_dir_path + "/weight_chunked_" + str(i) + ".npy"
        weight_chunked = np.load(path)

        ratio_close, ratio_equal = check_close_equal(weight, weight_chunked)
        all_equal = all_equal * ratio_equal
        print_close_equal("Weight layer {}".format(i), ratio_close, ratio_equal)

    path = test_dir_path + "/value.npy"
    write_equal(all_equal, path)


if __name__ == "__main__":
    test_sage_linear_adam()
